src/dataflow/src/sink/tail.rs

```diff
@@ -135,9 +145,11 @@ fn tail<G>(
                 }
             }
 
-            // Emit data if configured, otherwise it is an error to have data to send.
-            if let Some(tail_protocol) = &mut tail_protocol {
-                tail_protocol.send_rows(results);
+            if active_worker {
+                // Emit data if configured, otherwise it is an error to have data to send.
```

Comment:
Is it possible to have `active_worker && tail_protocol_handle is None`? This seems odd to me and if it's not the case, we wouldn't actually need the `active_worker` variable, but only decide to do work based on that the handle is `Some(_)`. But, I might be missing the relevant code!

Reply:
> Is it possible to have active_worker && tail_protocol_handle is None? 

Yes. After shutdown is initiated, this code runs:

```
        // Inform the coordinator that we have been dropped,
        // and destroy the tail protocol so the sink operator
        // can't send spurious messages while shutting down.
        Some(Box::new(scopeguard::guard((), move |_| {
            if let Some(tail_protocol_handle) = tail_protocol_weak.upgrade() {
                std::mem::drop(tail_protocol_handle.borrow_mut().take())
            }
        })))
```

which causes the tail protocol to be dropped (and both this scopeguard and the operator are sharing the same tail protocol object, behind an Rc<RefCell<_>>), but the dataflow operator can still be getting progress messages.